src/hyperspectral/hyperspectral/decorrelation/mif.py

```diff
@@ -0,0 +1,343 @@
+        kern = torch.unsqueeze(tkern, 0)
+        kern = torch.unsqueeze(tkern, 0)
+        imf_n, _ = imf_torch(f, kern, τ, max_iters)
+        imfs.append(imf_n)
+        f = f - imf_n
+        if terminated_torch(f[0,0,:,:]):
+            if use_tqdm:
+                rng.update(max_imfs)
+                rng.close()
+            break
+
+    return imfs, f
+
+
+def mif(signal, χ=1.6, τ=0.001, max_iters=1000, max_imfs=25):
+    """
+    Multidimensional iterative filter
+
+    Arguments:
+        signal (np.array): The image to decorrelate
+        χ: Spherical radius mutiplier
+        τ: The error threshhold for IMF generation
+        max_iters: Maximum allowable iterations for IMF generation
+        max_imfs: Maximum allowable number of IMFs to generate
+
+    References:
+        Cicone, A., & Zhou, H. (2017).  Multidimensional iterative filtering method
+        for the decomposition of high–dimensional non–stationary signals. Numerical
+        Mathematics: Theory, Methods and Applications, 10(2), 278-298.
+    """
+    f = signal - np.mean(signal)
+    imfs = []
+
+    try:
+        from tqdm.autonotebook import tqdm
+        rng = tqdm(range(max_imfs))
+    except:
+        rng = range(max_imfs)
+
+    for i in rng:
+        k = spherical_radius(f, χ)
+        kernel = get_mask_2d(k)
+        imf_n, _ = imf(f, kernel, τ, max_iters)
+        imfs.append(imf_n)
+        f = f - imf_n
+        if terminated(f[0,0,:,:]):
+            if use_tqdm:
+                rng.update(max_imfs)
+                rng.close()
+            break
+
+    return imfs, f
+
+
+def mif_decorrelate(img, χ=1.6, τ=0.001, max_iters=1000, max_imfs=25, dev=None):
+    """
+    Decorrelate an image with MIF
+
+    Arguments:
+        img (np.array or torch.tensor): The image to decorrelate; if using torch, also set dev
+        χ: Spherical radius mutiplier
+        τ: The error threshhold for IMF generation
+        max_iters: Maximum allowable iterations for IMF generation
+        max_imfs: Maximum allowable number of IMFs to generate
+        dev: Torch device, if applicable
+
+    References:
+        Cicone, A., Liu, J., & Zhou, H. (2016). Hyperspectral chemical plume
+        detection algorithms based on multidimensional iterative filtering
+        decomposition. Philosophical Transactions of the Royal Society A:
+        Mathematical, Physical and Engineering Sciences, 374(2065), 20150196.
+    """
+    rng = range(img.shape[0])
+    try:
+        from tqdm.autonotebook import tqdm
+        rng = tqdm(rng)
+    except: pass
+
+    if torch_available and dev is not None:
+        do_mif = partial(mif_torch, dev)
+        use_torch = True
+    else:
+        do_mif = mif
+        use_torch = False
+
+    decorrelated = []
+    for i in rng:
+        _, r = do_mif(img[i,:,:], χ, τ, max_iters, max_imfs)
+        resid = r.detach().cpu().numpy() if use_torch else r
+        decorrelated.append(img[i,:,:] - resid[0,0,:,:])
```

Comment:
I can't find `do_mif` so I don't know exactly what is in `resid[0, 0, :, :]`, but if it is just the last `imf` -- and if results are not satisfactory -- then it might make sense to experiment with subtracting more than just the last one (or changing `max_imfs`, &c).